### PR TITLE
Bindable picker

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

This PR includes and replace #296 by @joacar

### Bugs Fixed ###

None

### API Changes ###

Added:

	public static readonly BindableProperty Picker.ItemsSourceProperty;
	public static readonly BindableProperty Picker.SelectedItemProperty;
	public BindingBase Picker.ItemDisplayBinding { get; set;}
	public IList ItemsSource { get; set; }
	public object SelectedItem { get; set; }

### Behavioral Changes ###

None

### PR Checklist ###

- [X] Has tests (if omitted, state reason in description)
- [X] Rebased on top of master at time of PR
- [X] Changes adhere to coding standard
- [X] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/Picker.cs [hunk 1]
```diff
@@ -8,63 +9,325 @@ namespace Xamarin.Forms
...
+			BindableProperty.Create("Display", typeof(string), typeof(Picker), default(string));
+
+		string GetDisplayMember(object item)
+		{
+			if (ItemDisplayBinding == null)
+				return item.ToString();
+
+			ItemDisplayBinding.Apply(item, this, s_displayProperty);
+			ItemDisplayBinding.Unapply();
+			return (string)GetValue(s_displayProperty);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: See how wicked this is ?

**Comment by joacar** [comment]
How is the property `ItemDisplayBinding` used? Similar to "`DisplayMemberPath`" (aligns well with the same concept in WPF controls) - i.e. finding a property to use as display value on the bounded object(s)?

File: Xamarin.Forms.Core/Picker.cs (hunk 1 above)
Comment (joacar) [inline Xamarin.Forms.Core/Picker.cs]: An alternative way of doing it is the below code (which allows for nested properties)

````c#
// Find the property by walking the display member path to find any nested properties
            var propertyPathParts = DisplayMemberPath.Split('.');
            object propertyValue = item;
            foreach (var propertyPathPart in propertyPathParts)
            {
                var propInfo = propertyValue.GetType().GetTypeInfo().GetDeclaredProperty(propertyPathPart);
                if (propInfo == null)
                {
                    throw new ArgumentException($"No property '{propertyPathPart}' was found on '{propertyValue.GetType().FullName}'");
                }
                propertyValue = propInfo.GetValue(propertyValue);
            }
            if (propertyValue == null)
            {
                throw new ArgumentException($"No property '{DisplayMemberPath}' was found on '{item.GetType().FullName}'");
            }
            return propertyValue as string;
````

Don't know how it compares to `BindingBase` in terms of effeciency or ease to use from XAML though

Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Picker.cs]: Binding has the following advantages over your fully working and correct code:
- supports converters natively (no need for converter or Func properties)
- paths can contains indexers (I don't see this used often in Picker, but who knows)
- Bindings can be compiled while used from Xaml
- Bindings are fully tested

Comment (joacar) [inline Xamarin.Forms.Core/Picker.cs]: Awesome! Gonna try find me some good resources on `BindingBase` - seems like I overlooked it :)

**Comment by StephaneDelcroix** [comment]
@joacar this is how to use the `ItemDisplayBinding`:

     var picker = new Picker
     {
        ItemDisplayBinding = new Binding("Name"),
        ItemsSource = items
     };

or from Xaml

    <Picker ItemDisplayBinding="{Binding Name}" />

Xaml is the reason why the property is not Bindable. Otherwise the parser wouldn't now if it has to set the property, or apply the Binding.

**Comment by joacar** [comment]
Perhaps change all explicit type conversions of `LockableObservableListWrapper` on `Items` and introduce a private field

````
LockableObservableListWrapper lockableList;
...
public Picker()
{
    ....
    lockableList =(LockableObservableListWrapper)Items;
}
````

**Comment by StephaneDelcroix** [comment]
direct casts are free

File: Xamarin.Forms.Core/Picker.cs [hunk 2]
```diff
@@ -8,63 +9,325 @@ namespace Xamarin.Forms
...
 
+		BindingBase _itemDisplayBinding;
+		public BindingBase ItemDisplayBinding {
+			get { return _itemDisplayBinding; }
+			set {
+				if (_itemDisplayBinding == value)
+					return;
+
+				OnPropertyChanging();
+				var oldValue = value;
```
Comment (lthib) [inline Xamarin.Forms.Core/Picker.cs]: Noticing (long after the fact) oldValue is in fact assigned the newly set value. Won't have any impact the way the code doesn't make use of these values in OnItemDisplayBindingChanged, but may cause future issues.

**Comment by xrkolovos** [comment]
Do you know when 2.3.4 is going to be stable?

**Comment by bill2004158** [comment]
any idea why I got this error when build:
No property, bindable property, or event found for 'ItemsSource'

ps:
I am using id="Xamarin.Forms" version="2.3.4.231"

**Comment by Magic73** [comment]
I'm using 2.4.. and it works.

**Comment by JohnGalt1717** [comment]
Thank you for taking a stab at this but:

This should have Value, DisplayTextField and ValueField properties that are bindable along with ItemsSource.

It should allow the itemsSource to be bound to any collection of objects and then ValueField="Id" DisplayTextField="Name" Value={Binding ResultId}

Then add an ItemTemplate that allows you to customize the layout based on the entire object if you want.

There is absolutely no other correct way of doing a drop down control. All other ways are ridiculous wastes of time that completely fail to understand the actual problem.

Lists are always used to provide a list of items that you wish to display and then get the identifier value. If you want the entire thing then set the ValueField to null so that it returns the entire object.  This is directly related to how databases are setup with relationships.

It took Microsoft until WPF to finally make a usable Combobox/Dropdown that would actually work in 95% of cases. I'm shocked that Xamarin didn't learn the lesson the first time around. Having to use a 3rd party control for such a basic concept is unfathomable.

**Comment by davidbuckleyni** [comment]
Would be great if this could support material in that a small label while its not picked or some kind of warning.

**Pull request merged**